### ISSUE TITLE
AUDIT: Various kinds of observations

### ISSUE THREAD

**Issue by asalzmann** [Stride-Labs/stride#424]
---
finding:
  id: IF-STRIDE-OBSERVATIONS
  title: Various kinds of observations
  severity: Informational
  impact: None
  exploitability: None
  type: Implementation
  issue: 
  status: Unresolved
---

# Description

- **There are no invariants defined in Stride codebase**. In the context of the Cosmos SDK, an Invariant is a function that checks for a particular invariant at the end of each block. These functions are useful to detect bugs early on and act upon them to limit their potential consequences (e.g. by halting the chain). They are also useful in the development process of the application to detect bugs via simulations.
- **Simulation tests not implemented**. The Cosmos SDK offers a full-fledged simulation framework to fuzz-test every message defined by a module. Although Stride modules contain scaffolded skeletons for simulation messages, they are not implemented (e.g. LiquidStake msg):  
```go
    func SimulateMsgLiquidStake(
  ak types.AccountKeeper,
  bk types.BankKeeper,
  k keeper.Keeper,
    ) simtypes.Operation {
  return func(r *rand.Rand, app *baseapp.BaseApp, ctx sdk.Context, accs []simtypes.Account, chainID string,
      ) (simtypes.OperationMsg, []simtypes.FutureOperation, error) {
    simAccount, _ := simtypes.RandomAcc(r, accs)
    msg := &types.MsgLiquidStake{
      Creator: simAccount.Address.String(),
    }

    // TODO: Handling the LiquidStake simulation

    return simtypes.NoOpMsg(types.ModuleName, msg.Type(), "LiquidStake simulation not implemented"), nil, nil
      }
    }
```
- `HostZoneUnbonding`'s status `EXIT_TRANSFER_QUEUE` was set twice, [before](https://github.com/Stride-Labs/stride/blob/v2.0.3/x/stakeibc/keeper/icacallbacks_undelegate.go#L95) and [after](https://github.com/Stride-Labs/stride/blob/v2.0.3/x/stakeibc/keeper/icacallbacks_undelegate.go#L106) burning `stTokens` during `UndelegateCallback`.

- `SetWithdrawalAddressOnHost` is being called [on every epoch](https://github.com/Stride-Labs/stride/blob/v2.0.3/x/stakeibc/keeper/hooks.go#L75-L91).
- Flags: `HostZoneUnbonding_UNBONDING_IN_PROGRESS` and `HostZoneUnbonding_EXIT_TRANSFER_IN_PROGRESS` seem to be unused.

# Problem Scenarios

# Recommendation
-  Think about adding some invariants. Describe and check conditions on Stride which should be fulfilled at every block execution or on every epoch.
- If there is a need for fuzzy testing, you could utilize cosmos sdk simulator by implementing messages of interest. These are simulated with random field values. The sender of the operation is also assigned randomly. The simulator can test parameter changes at random as well. Finally, you could create a randomized genesis file.
- Consider representing `amount` fields with string type (if their type `int64` or `uint64`). The way protobuf json marshaling works, it encodes the `uint64` as a `string`. This also allows for larger amounts such as `uint256` to be sent across chains. An example of this approach can be found in [proto/ibc/applications/transfer/v2/packet.proto](https://github.com/cosmos/ibc-go/blob/main/proto/ibc/applications/transfer/v2/packet.proto):

```go
message FungibleTokenPacketData {
  // the token denomination to be transferred
  string denom = 1;
  // the token amount to be transferred
  string amount = 2;
  // the sender address
  string sender = 3;
  // the recipient address on the destination chain
  string receiver = 4;
}
```

- Set `HostZoneUnbonding`'s status `EXIT_TRANSFER_QUEUE` either before or after burning `stTokens` during `UndelegateCallback`.
- `SetWithdrawalAddressOnHost` should not be called on every epoch.
- Define only necessary flags for `HostZoneUnbonding_Status`. Probably `HostZoneUnbonding_UNBONDING_IN_PROGRESS` and `HostZoneUnbonding_EXIT_TRANSFER_IN_PROGRESS` should be removed.

**Comment by sampocs** [comment]
@asalzmann do you know what they mean by the last bullet
> Define only necessary flags for HostZoneUnbonding_Status. Probably HostZoneUnbonding_UNBONDING_IN_PROGRESS and HostZoneUnbonding_EXIT_TRANSFER_IN_PROGRESS should be removed.

I think those are used (implicitly) and should not be removed. They're also useful for our monitoring

**Comment by asalzmann** [comment]
Yes, it's used implicitly. For example, here https://github.com/Stride-Labs/stride/blob/0dbbbd867ffad5b331d09c155dca53a3f581ad5c/x/stakeibc/keeper/unbonding_records.go#L295

If we didn't have another state, we wouldn't know which records to process. It's also useful for monitoring. Removing that comment here and will ping informal/

**Comment by ThanhNhann** [comment]
Yeah I think so @sampocs , I have checked it and commented I this https://github.com/Stride-Labs/stride/issues/525
